### PR TITLE
[r+] Add enum discriminants to the reference.

### LINKED ISSUE #15755: Document integer enum discriminators in the manual
The tutorial documents how to set the discriminator for a C-style enum (`enum Foo { Bar = 123 }`) and how to cast such an enum to a numeric type (`Bar as uint`).  However, the language manual does not mention these features at all.

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #15755

### CODE REVIEW

File: src/doc/reference.md [hunk 1]
```diff
@@ -1413,6 +1413,21 @@ a = Animal::Cat { name: "Spotty".to_string(), weight: 2.7 };
 In this example, `Cat` is a _struct-like enum variant_,
 whereas `Dog` is simply called an enum variant.
 
+Enums can have a discriminator:
```
Comment (sfackler) [inline src/doc/reference.md]: Enums always have a discriminant, this syntax just lets you specify exactly what it is. They'll default to 0, 1, 2, ...

**Comment by steveklabnik** [comment]
@sfackler addressed

File: src/doc/reference.md [hunk 2]
```diff
@@ -1413,6 +1413,24 @@ a = Animal::Cat { name: "Spotty".to_string(), weight: 2.7 };
...
+```
+enum Foo {
+    Bar = 123,
+}
+```
+
+If a discriminant isn't assigned, they start at zero, and add one for each
+variant, in order.
+
+You can cast an enum to get this value:
```
Comment (sfackler) [inline src/doc/reference.md]: One last nit: you can only do this if the enum is C-like, i.e. none of the variants have any data attached. For example, you can't do `None as u32`.

Comment (steveklabnik) [inline src/doc/reference.md]: >  C-like, i.e. none of the variants have any data attached. 

By "C-like" you mean none like `Foo(i32)`, yes?

Comment (sfackler) [inline src/doc/reference.md]: Yep! PR #20907 adds an intrinsic to extract he discriminant from any enum, but its only intended to be used as an optimization for things like `PartialEq` implementations.

**Comment by nikomatsakis** [comment]
r+ with @sfackler 's nit addressed

**Comment by steveklabnik** [comment]
Doing this manually as part of https://github.com/rust-lang/rust/pull/21300

**Pull request closed without merging**